core/trino-spi/src/main/java/io/trino/spi/block/Block.java

```diff
@@ -327,4 +327,10 @@ default Block getLoadedBlock()
     {
         return Collections.emptyList();
     }
+
+    /**
+     * Returns a block that has an appended null at the end, no matter if the original block has null or not.
+     * The original block won't be modified.
+     */
+    Block appendNull();
```

Comment:
This seems like a builder responsibility. I think this breaks logical separation between `Block` and it's builder. However, I'm not sure if long term we will still use builders as we aim for more batch oriented approach.
cc @martint

Reply:
@sopel39 good point, I'm a bit concerned about this too. Would it be a little bit better to name this `Block newBlockWithAppendedNull()`, and explicitly throwing in that method for all builders?

Reply:
I'm not sure how it's being used. I wonder if it shouldn't be:
1. create new Block builder for type
2. Have new method like `writePositionsTo` which would write sequence of positions to block builder
3. append null
4. construct new block.

In any case, this operation seems rather expensive to just append single position. Can it be avoided?

Reply:
the usage can be summarized as: being able to build a dictionary block with nulls, out of any source block that may not have nulls - while avoiding copies as much as possible. For the dictionary to represent nulls, the source must have nulls. The 4-step approach above will do more copies, whereas the current approach will be able to reuse things.

Reply:
> The 4-step approach above will do more copies, whereas the current approach will be able to reuse things.

why its' the case?

IIUC `appendNull` will have to copy arrays in primitive block, but also arrays in `DictionaryBlock`

Reply:
in primitive blocks, only isnull array may need to be copied if values is already a slice of a bigger thing. In arrray, map, row block, and variable width block, data doesn't have to be copied. For dictionary block, we may only end up copying ids if input has null (otherwise it depends on what the source block is). so looks like having this method in `Block` allows us to optimize based on knowledge of internal structures case-by-case. I'm not sure how we'd be able to accomplish that with `writePositionsTo` in BlockBuilder.

Reply:
How many times this new method is used per block? Assuming `O(n)` worst complexity, if it's being used multiple times, then it could blow up Block computations?

If you need to have a null, isn't it better to just output single position null block?

cc @dain

Reply:
this is at-most once per input block.

> If you need to have a null, isn't it better to just output single position null block?

we'd still want to use other non-null elements in the dictionary block. (Other alternative would be to change DictionaryBlock contract to assume some id like `-1` as null, haven't looked into what the implications of that would be)

Reply:
> we'd still want to use other non-null elements in the dictionary block. 

why these can be 2 separate blocks (one with dictionary and other with single null value)?

Reply:
> Is it because null positions are interleaved with non-null dictionary values?

right, if we're `unnesting (col1, col2)`, and build output page with say 3 input rows.

```markDown
| col0 | col1            | col2            |
|------|-----------------|-----------------|
| a    | ["b", "c"]      | ["d", "e", "f"] |
| g    | ["h", "i", "j"] | ["k"]           |
| l    | ["m", "n"]      | ["o", "p", "q"] |
```

for output block of unnested `col1`, we need `b, c, null, h, i, j, m, n, null`. For `col2`, we need `d, e, f, k, null, null, o, p, q`. So null positions are interleaved in dictionary.